bin/ember

```diff
@@ -25,6 +25,40 @@ try {
 
 logger.info('Resolved "ember-cli" to %j', projectLocalCli);
 
+if (!projectLocalCli && !process.versions.pnp) {
+  // If we haven't found `ember-cli` locally it might be installed though
+  // yarn plug'n'play. Let's check if an `.pnp.js` file exists and if so,
+  // use it to find `ember-cli` in the yarn cache and setup yarn PnP support
+  // for the following `require()` calls.
+  var findUp = require('find-up');
+
+  logger.info('Checking for ".pnp.js" file from %j...', basedir);
+  var pnpFile = findUp.sync('.pnp.js', { cwd: basedir });
+  logger.info('-> pnpFile: %j', pnpFile);
+  if (pnpFile) {
+    logger.info('Loading Yarn PnP API...');
+    var pnp = require(pnpFile);
+
+    logger.info('Setting up Yarn PnP compatibility layer...');
+    pnp.setupCompatibilityLayer();
+
+    logger.info('Setting up Yarn PnP Node.js hooks...');
+    pnp.setup();
+
+    // We add `--require ${pnpFile}` to the NODE_OPTIONS environment variable
+    // here to ensure that any `child_process` calls will use Yarn PnP mode too
+    logger.info('Adding ' + pnpFile + ' to NODE_OPTIONS...');
+    process.env.NODE_OPTIONS = (process.env.NODE_OPTIONS || '') + ' --require ' + pnpFile;
+
+    logger.info('Resolving "ember-cli" through Yarn PnP...');
+    try {
+      projectLocalCli = pnp.resolveToUnqualified('ember-cli', basedir, { considerBuiltins: false });
+    } catch(ignored) {}
```

Comment:
Should this catch all exceptions, or a specific one? In general, I prefer being explicit, so as it not hide unexpected issues.

Reply:
Ya, definitely possible to check against a list of PnP error codes. Some of them are used here:

https://github.com/ember-cli/ember-cli-version-checker/blob/8c3a97d0bb23352ac31e69f754111ef004ff449e/src/npm-dependency-version-checker.js#L11-L14

But we may end up having others, should review the code to confirm we hit them all. You can look for `makeError` usage in https://github.com/yarnpkg/yarn/blob/master/src/util/generate-pnp-map-api.tpl.js to get a full list.